### PR TITLE
feat: detect weak crypto algorithm

### PR DESCRIPTION
**Pull request by tony-go**
## Context

This PR resolves issue #25 

## Improvements 🚀 

At the moment we detect:

| Algorithm | Code              | import 'crypto' |
|-----------|-------------------|-----------------|
| `md5`     | createHash('md5') |  ✅              |
| `md5`     | crypto.createHash('md5') |  ✅              |

## Questions

- Should we make `import 'crypto'` be required to trigger this warning?
- Should I add other weak crypto algorithms in this PR? (like `sha1`, but TBH I don't know other ones 🙈)?

### CODE REVIEW

**Review by Kawacrepe — COMMENTED** [review]
Lgtm! My only question is: shouldn't we add a valid test case ? where the cryptoHash algorithm used is something like sha256 etc..?

**Comment by tony-go** [comment]
> Lgtm! My only question is: shouldn't we add a valid test case ? where the cryptoHash algorithm used is something like sha256 etc..?

Very good idea ^^ I'll add it.

**Review by Kawacrepe — APPROVED** [review]

**Comment by tony-go** [comment]
@fraxken do you prefer to wait for the tracer feature?

**Comment by fraxken** [comment]
@tony-go No i guess this is ok to merge. For weak hash i have found this list
- MD2, MD4, MD5, RIPEMD-160, and SHA-1

SHA1 is supported by Node.js (maybe we have to check for the other ones).

File: src/probes/isWeakCrypto.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    node.callee.property.name === "createHash";
+
+  return [isSimpleIdentifier || isMemberExpression];
+}
+
+function main(node, { analysis }) {
+  const arg = node.arguments.at(0);
+  if (arg.value === "md5" && analysis.dependencies.has("crypto")) {
+    analysis.addWarning(warnings.weakCrypto, "md5", node.loc);
+  }
```
Comment (fraxken) [inline src/probes/isWeakCrypto.js]: Make the code dynamic to check multiple hash algorithms. The list of vulnerable hashes could be stored in an ES6 Set.

File: index.js [hunk 2]
```diff
@@ -128,6 +128,11 @@ export const warnings = Object.freeze({
     code: "obfuscated-code",
     i18n: "sast_warnings.obfuscated_code",
     severity: "Critical"
+  },
+  weakCrypto: {
+    code: "weak-crypto",
+    i18n: "sast_warnings.weak_crypto",
+    severity: "Information"
```
Comment (fraxken) [inline index.js]: Maybe we should add another property `experimental: true` here ? So other tools like CLI may use this to show an experimental note.

**Review by fraxken — CHANGES_REQUESTED** [review]

File: src/probes/isWeakCrypto.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
+// Internal Dependencies
+import { warnings } from "../constants.js";
+
+// Constants
+const weakAlgorithms = ["md5", "sha1", "ripemd160", "md4", "md2"];
```
Comment (fraxken) [inline src/probes/isWeakCrypto.js]: `new Set(["md5", "sha1", "ripemd160", "md4", "md2"])` no ?

File: src/probes/isWeakCrypto.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+  return [isSimpleIdentifier || isMemberExpression];
+}
+
+function main(node, { analysis }) {
+  const arg = node.arguments.at(0);
+  const isCryptoImported = analysis.dependencies.has("crypto");
+
+  if (
+    weakAlgorithms.includes(arg.value) &&
```
Comment (fraxken) [inline src/probes/isWeakCrypto.js]:
```js
weakAlgorithms.has(arg.value)
```

With an ES6 Set

File: src/probes/isWeakCrypto.js (hunk 3 above)
Comment (tony-go) [reply src/probes/isWeakCrypto.js]: Shame on me! this is the vim effect!

File: src/probes/isWeakCrypto.js (hunk 4 above)
Comment (tony-go) [reply src/probes/isWeakCrypto.js]: 🙈

**Comment by fraxken** [comment]
@all-contributors please add @tony-go for code,tests,docs

**Comment by fraxken** [comment]
@all-contributors please add @tony for code, doc, test

**Pull request merged**